mlc_llm/transform/smoothquant.py

```diff
@@ -0,0 +1,245 @@
+                self.pattern = is_op("relax.matmul")(self.lhs_sm, self.permute)
+            
+            def transform(self) -> tvm.IRModule:
+                for gv, func in self.mod.functions.items():
+                    if not isinstance(func, relax.Function):
+                        continue
+                    self.var2val = tvm.relax.analysis.get_var2val(func)
+                    self.profile_points = []
+                    self.params_to_remove = []
+                    updated_func = self.visit_expr(func)
+                    updated_func = remove_all_unused(updated_func)
+                    self.builder_.update_func(gv, updated_func)
+                return self.builder_.get()
+            
+            def visit_function_(self, f):
+                body = super().visit_expr(f.body)
+                new_params = [param for param in f.params if param not in self.params_to_remove]
+                return relax.Function(new_params, body, None, f.is_pure, f.attrs, f.span)
+            
+            def visit_seq_expr_(self, op: relax.SeqExpr) -> relax.Expr:
+                op = super().visit_seq_expr_(op)
+                if len(self.profile_points) != 0:
+                    new_body = relax.Tuple([op.body, relax.Tuple(self.profile_points)])
+                    return relax.SeqExpr(op.blocks, new_body, op.span)
+                return op
+
+            def visit_dataflow_block_(self, block: relax.DataflowBlock) -> relax.DataflowBlock:
+                self.builder_._begin_dataflow_block()
+                for binding in block.bindings:
+                    self.visit_binding(binding)
+                if len(self.profile_points) != 0:
+                    self.builder_.emit_output(self.profile_points)
+                return self.builder_._end_block()
+
+            def visit_call_(self, call: relax.Call) -> relax.Expr:
+                call = super().visit_call_(call)
+                matchings = self.pattern.extract_matched_expr(call, self.var2val)
+                if matchings:
+                    m_smq1 = matchings[self.lhs_sm]
+                    m_smq2 = matchings[self.rhs_sm]
+                    a_info = matchings[self.a_scale].struct_info
+                    w_info = matchings[self.w_scale].struct_info
+                    a_scale = tvm.runtime.ndarray.empty(a_info.shape, a_info.dtype, tvm.cpu())
+                    w_scale = tvm.runtime.ndarray.empty(w_info.shape, w_info.dtype, tvm.cpu())
+                    a_smq = self.builder_.emit(
+                        R.smooth(m_smq1.args[0], relax.Constant(a_scale), kind=1, mode="identity")
+                    )
+                    w_smq = self.builder_.emit(
+                        R.smooth(m_smq2.args[0], relax.Constant(w_scale), kind=2, mode="identity")
+                    )
+                    a_out = self.builder_.emit(R.absmax(a_smq, kind=1), "a_out")
+                    w_out = self.builder_.emit(R.absmax(w_smq, kind=2), "w_out")
```

Comment:
Do we need `R.absmax` at all, given that it is only used here and it is just legalized anyway?

Reply:
I will answer this question in the following comment.